ParameterState: Add IsChildOriginatedChange to ParameterChangedEventArgs

ScarletKuro opened the pull request:
## Description
To better understand, here is the GIF:
![state](https://github.com/user-attachments/assets/34601896-cdc7-4dbd-a2e3-4a3616afc259)

As you can see from the GIF, with `WithChangeHandler`, we can now distinguish whether the state was updated by the child (via `SetValueAsync`) and propagated to the parent, causing the parent to update the related [Parameter], or if the change was done solely by the parent.

### Child component:
```razor
<div class="child-div">
    Child Component
    <MudButton id="childBtn" OnClick="OnClickChildAsync">Click Child</MudButton>
</div>

@code {
    //... skipping code for 'RegisterParameter'
    [Parameter]
    public int Counter { get; set; } = 0;

    [Parameter]
    public EventCallback<int> CounterChanged { get; set; }

    private Task OnClickChildAsync() => _counterState.SetValueAsync(_counterState.Value + 1);
}
```

### Parent component:
```razor
<div class="parent-div">
    Parent Component
    <MudButton id="parentBtn" OnClick="OnClickParentAsync">Click Parent</MudButton>
    <ParameterStateChildComp @bind-Counter="Counter" />
</div>

@code {
    public int Counter { get; internal set; } = 1;

    private Task OnClickParentAsync()
    {
        Counter++;
        return Task.CompletedTask;
    }
}
```

If the parameter change originates from `OnClickChildAsync`, then `IsChildOriginatedChange` will be `true`. If the change is made by `OnClickParentAsync`, then `IsChildOriginatedChange` will be `false`.

As you can see, when the component is first opened, it already contains:
```
Counter: 0 => 1 by Parent
```
This is expected because the default `Counter` value for the child component is `0`, but the parent overrides the initial value to `1` during initialization. Therefore, we get the correct data. If the `Parent` had `Counter { get; internal set; } = 0`, then no change would have occurred.

Sometimes, I find it necessary to understand where the change originated from. For example, for validation, if I build it around the change handler of the parameter. If the change originates from the child, I can trigger the validation right away after `SetValueAsync` so that when I receive the change handler, I don’t want to run it again. But if the value was changed by the user of my component, then I want to run the validation on the change handler.
Of course, if we didn’t run the validation on the child side after `SetValueAsync` and instead waited for the parent to change it and strictly validated it in the change handler, then it wouldn't be required. However, this is just one use case, and sometimes you have to work with legacy code that has already established certain logic. In those situations, you may not want to rewrite everything entirely, as in the case of ColorPicker (https://github.com/MudBlazor/MudBlazor/pull/10357). So, I think this information will not hurt in any case.

## How Has This Been Tested?
<!-- All PRs should implement unit tests if possible. -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

ScarletKuro committed:
ParameterState: Add IsChildOriginatedChange

ScarletKuro committed:
nit

ScarletKuro committed:
Add test for this feature

ScarletKuro commented:
Alternative names that we could use:
- `IsPropagatedFromChild`
- `IsChildToParentPropagation`
- `WasChangedByChild`
- `IsChildInitiatedChange`
- `IsChildDrivenUpdate`
- `IsUpdateFromChild`
- `WasPropagatedUpward`

But I kind of like the current one.

ScarletKuro committed:
Remove not used namespaces

henon commented on src/MudBlazor/State/ParameterStateInternalOfT.cs:
```diff
@@ -101,6 +102,7 @@ public void OnParametersSet()
         var currentParameterValue = _getParameterValueFunc();
         if (!_comparer.Equals(_lastValue, currentParameterValue))
         {
+            _isChildOriginatedChange = _comparer.Equals(_value, currentParameterValue);
```
How does this line do the magic of detecting the origin?

ScarletKuro replied:
When the value originates from the child, the `_value` is updated internally by the `SetValueAsync` method
https://github.com/MudBlazor/MudBlazor/blob/c35e50cf52ab8a6fa5ddf93abbacafa3f7ba5e65/src/MudBlazor/State/ParameterStateInternalOfT.cs#L79

The synchronization happens at the `OnParametersSet` method (in case it was only updated by the parent only). Therefore, if we compare the `_value` with the `currentParameterValue` (the value set by the parent) before 
https://github.com/MudBlazor/MudBlazor/blob/c35e50cf52ab8a6fa5ddf93abbacafa3f7ba5e65/src/MudBlazor/State/ParameterStateInternalOfT.cs#L106
We can determine the origin of the value.  

Theoretically, we could set a flag in `SetValueAsync` as well and reset it on read, but I find the current approach more reliable, especially in cases where something might get unexpectedly overwritten. For this reason, I believe the `OnParametersSet` lifecycle method is the best option here. I also asked Copilot how it would handle this without giving any context about `ParameterState` of its current implementation, and it also suggested similar approach with `OnParametersSet`.

henon commented:
If you look at a hierarchy `A -> B -> C` where A contains B and B contains C  (and of course there are two-way bindings between A and B and B and C), aren't there three possible origins for a parameter change in B? 
1. Changed by Parent (A is the origin)
2. Changed by Self (B is the origin)
3. Changed by Child (C is the origin)

But I thought that the change handler is not invoked if the change was done by self via `SetValueAsync` so I guess the second one should be irrelevant?

henon reviewed (approved):
In any case a useful tool. I'd just like to understand how it works ;)

ScarletKuro commented:
![statehierarchy](https://github.com/user-attachments/assets/8fbdb18f-812b-480a-9986-54b4bf6e61d2)

I updated the gif and moved the "parameter change" to their correct color, as in header gif I didn't bother with it as it wasnt that important with two components, but with nesting it's easier to understand where it the change handler belongs to

ScarletKuro committed:
Update test to include nesting

ScarletKuro commented:
Update test to include the nesting scenario below to fixate this behavior.

ScarletKuro committed:
Format css

henon commented:
So if we go back to your original example with just parent and child, how I understand it now is this:

When the child updates the counter it first sets it in its own parameter state then propagates the value up to the parent via change event. The parent will set the parameter on the child again due to the two-way binding. Then in the child the logic in `OnParameterSet` sets the flag `IsChildOriginatedChage` to true because the internal value and the incoming value are the same.

If I understand this correctly now then I think the flag is not named well. For the external observer in this configuration, the logic happens in the child. From the childs perspective, where the flag is computed it is in fact distinguishing an internally originated change from an external (parent-originated) change. 

The problem with the current name becomes obvious in the `A, B, C` example when we look at `B`. It would falsely suggest that a change with `IsChildOriginatedChange` originated from `C`. 

Maybe I am still not understanding correctly? But if I do then a name such as `IsSelfOriginatedChange` or similar would be much clearer.

ScarletKuro commented:
> When the child updates the counter it first sets it in its own parameter state then propagates the value up to the parent via change event.

Yes.

> The parent will set the parameter on the child again due to the two-way binding.

Yes, but remove word "again". The child only updated the internal state parameter state, and parent updates the [Parameter] of the child itself now because of the two way binding https://github.com/dotnet/aspnetcore/discussions/49187#discussioncomment-6372174.

> Maybe I am still not understanding correctly? But if I do then a name such as IsSelfOriginatedChange or IsOwnChange or similar would be much clearer.

Sure, `IsSelfOriginatedChange` sounds good to me.
